File: Tennis.Play/PlayTournament.cs
using System;
using Tennis.Logic;
using System.Collections.Generic;

namespace Tennis.Play
{
	public class PlayTournament : IPlayTournament
	{
		private readonly List<IPlayMatch> playMatches;
		private List<MatchScore> matchScores;
		
		public PlayTournament (List<IPlayMatch> playMatches)
		{
			if (playMatches == null)
			{
				throw new ArgumentNullException("sides");
			}
			
			if (playMatches.Count == 0)
			{
				throw new ArgumentException("sides");
			}
			
			this.playMatches = playMatches;
		}
		
		public ISide Play()
		{
			ISide champion = null;
			matchScores = new List<MatchScore>();
			
			PlayMatches(playMatches, ref champion);
			
			return champion;
		}
		
		public List<MatchScore> GetMatchScores()
		{
			return this.matchScores;
		}
		
		private void PlayMatches(List<IPlayMatch> matches, ref ISide champion)
		{
			var nextRound = new List<IPlayMatch>();
			var winningSides = new List<ISide>();
			
			if (matches.Count == 1)
			{
				champion = matches[0].Play();
				AddMatchScore(matches[0]);
				return;
			}
			
			foreach(var match in matches)
			{
				var winner = match.Play();
				AddMatchScore(match);
				winningSides.Add(winner);
			}
			
			for (int i = 0; i < winningSides.Count; i = i + 2) 
			{
				if (i == winningSides.Count - 1 && winningSides.Count % 2 == 1)
				{
					nextRound.Add(new PlayMatch(winningSides[i], winningSides[i]));
				}
				else
				{
					nextRound.Add(new PlayMatch(winningSides[i], winningSides[i+1]));
				}
			}
			
			PlayMatches(nextRound, ref champion);	
		}
		
		private void AddMatchScore(IPlayMatch playMatch)
		{
			matchScores.Add(new MatchScore() 
				{ SetScores = playMatch.GetSetScores(), 
					SideOne = playMatch.SideOne, 
					SideTwo = playMatch.SideTwo });
		}
	}
}

File: Tennis.FSharp.Play/PlayTournament.fs
namespace Tennis.FSharp.Play
    
    open Tennis.FSharp.Logic
    open FSharpx

    type PlayTournament(matches: seq<IPlayMatch>) =
        let mutable matches = matches
        let mutable matchList = []              

        member private this.PlayTheMatch (aMatch: IPlayMatch) =
            let winner = aMatch.Play()
            let matchScore = new MatchScore()
            matchScore.SetScores <- aMatch.GetSetScores()
            matchList <- matchScore :: matchList
            winner
        

        member private this.PlayRound matches =
            [
                for aMatch in matches do
                    yield this.PlayTheMatch aMatch
            ]
                

        member private this.MatchUp (sides: seq<ISide>) =
            let len = Seq.length sides
            let med = len / 2
            let firstHalf = Seq.take med sides
            let lastHalf = Seq.skip med sides

            Seq.zip firstHalf lastHalf
                |> Seq.map (fun (first, second) -> new PlayMatch(new PlaySet(new PlayGame(new DetermineWinner(first, second))), first, second))
                :?> seq<IPlayMatch>

        interface IPlayTournament with
            member this.Play() = 
                let mutable currentMatch = Seq.head matches;
                let mutable tournamentWinner = currentMatch.SideOne
                let mutable continueLooping = true
                let mutable roundMatches = matches

                while continueLooping do                  
                    if Seq.length roundMatches = 1 then
                        tournamentWinner <- this.PlayTheMatch (Seq.head roundMatches)
                        continueLooping <- false
                    else
                        let sides = this.PlayRound roundMatches
                        let newMatches = this.MatchUp sides
                        roundMatches <- newMatches
                tournamentWinner

            member this.GetMatchScores() =
                List.rev matchList |> List.toSeq
